# Bank Marketing

### Install Requirements

In [0]:
! pip install memory_profiler

In [0]:
%load_ext memory_profiler

In [0]:
! apt-get install default-jre
! java -version
! pip install h2o

Reading package lists... Done
Building dependency tree       
Reading state information... Done
default-jre is already the newest version (2:1.10-63ubuntu1~02).
0 upgraded, 0 newly installed, 0 to remove and 4 not upgraded.
openjdk version "10.0.2" 2018-07-17
OpenJDK Runtime Environment (build 10.0.2+13-Ubuntu-1ubuntu0.18.04.4)
OpenJDK 64-Bit Server VM (build 10.0.2+13-Ubuntu-1ubuntu0.18.04.4, mixed mode)


### Import Libraries

In [0]:
from sklearn.preprocessing import MinMaxScaler
from h2o.estimators.glm import H2OGeneralizedLinearEstimator
from h2o.estimators.gbm import H2OGradientBoostingEstimator
from h2o.estimators.random_forest import H2ORandomForestEstimator
import h2o
import os

# Pretty Display of Variables
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [ ]:
def process_bank_data(ratios=[0.8, 0.1]):
    """ Downloads the banking dataset, preprocess the data and splits it.
    Args:
        ratio: Split ratio. Default is 0.8
    
    Returns:
        train, test and sample dataset
    
    """
    
    user = "fazilbtopal"
    key = "a01ead977f55d872c4deeadb0f173aa1"

    if '.kaggle' not in os.listdir('/root'):
        !mkdir ~/.kaggle
    !touch /root/.kaggle/kaggle.json
    !chmod 666 /root/.kaggle/kaggle.json
    with open('/root/.kaggle/kaggle.json', 'w') as f:
        f.write('{"username":"%s","key":"%s"}' % (user, key))
    !chmod 600 /root/.kaggle/kaggle.json

    print('Downloading data from web..')
    ! kaggle datasets download -d sonujha090/bank-marketing
    ! unzip -qq /content/bank-marketing.zip
    
    print('Reading the dataset..')
    # Load the data set.
    bank_df = h2o.import_file("/content/bank-full.csv")
    bank_sample_df = h2o.import_file("/content/bank.csv")


    print('Splitting train & test frames..')
    train, test, val = bank_df.split_frame(ratios=ratios, seed=1234)

    return train, test, sample

In [0]:
h2o.init(max_mem_size="4G")

Checking whether there is an H2O instance running at http://localhost:54321. connected.


--------------------------  ----------------------------------------
H2O cluster uptime:         2 hours 28 mins
H2O cluster timezone:       Etc/UTC
H2O data parsing timezone:  UTC
H2O cluster version:        3.22.1.3
H2O cluster version age:    10 days
H2O cluster name:           H2O_from_python_unknownUser_frr5wp
H2O cluster total nodes:    1
H2O cluster free memory:    3.998 Gb
H2O cluster total cores:    2
H2O cluster allowed cores:  2
H2O cluster status:         locked, healthy
H2O connection url:         http://localhost:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.6.7 final
--------------------------  ----------------------------------------

In [0]:
train, val, sample = process_bank_data()

In [0]:
train.head(10)

age job marital education default balance housing loan contact day month duration campaign pdays previous poutcome y 
 
 
 58 management married tertiary no 2143 yes no unknown 5 may 261 1 -1 0 unknown no 
 44 technician single secondary no 29 yes no unknown 5 may 151 1 -1 0 unknown no 
 33 entrepreneur married secondary no 2 yes yes unknown 5 may 76 1 -1 0 unknown no 
 33 unknown single unknown no 1 no no unknown 5 may 198 1 -1 0 unknown no 
 35 management married tertiary no 231 yes no unknown 5 may 139 1 -1 0 unknown no 
 28 management single tertiary no 447 yes yes unknown 5 may 217 1 -1 0 unknown no 
 42 entrepreneur divorced tertiary yes 2 yes no unknown 5 may 380 1 -1 0 unknown no 
 58 retired married primary no 121 yes no unknown 5 may 50 1 -1 0 unknown no 
 41 admin. divorced secondary no 270 yes no unknown 5 may 222 1 -1 0 unknown no 
 29 admin. single secondary no 390 yes no unknown 5 may 137 1 -1 0 unknown no

In [0]:
train.describe()

In [0]:
train.types

{'age': 'int',
 'balance': 'int',
 'campaign': 'int',
 'contact': 'enum',
 'day': 'int',
 'default': 'enum',
 'duration': 'int',
 'education': 'enum',
 'housing': 'enum',
 'job': 'enum',
 'loan': 'enum',
 'marital': 'enum',
 'month': 'enum',
 'pdays': 'int',
 'poutcome': 'enum',
 'previous': 'int',
 'y': 'enum'}

## Logistic Regression

In [0]:
feature_columns = train.col_names[:-1]
label = train.col_names[-1]

In [0]:
# Model parameters
model = H2OGeneralizedLinearEstimator(model_id="glm_v1", family="binomial", solver="AUTO")

In [0]:
model.train(x=feature_columns, y=label, training_frame=train, validation_frame=val)

glm Model Build progress: |███████████████████████████████████████████████| 100%


In [0]:
model

Model Details
H2OGeneralizedLinearEstimator :  Generalized Linear Modeling
Model Key:  glm_v1


ModelMetricsBinomialGLM: glm
** Reported on train data. **

MSE: 0.07126613766420797
RMSE: 0.2669571832039887
LogLoss: 0.23983725623041302
Null degrees of freedom: 36223
Residual degrees of freedom: 36182
Null deviance: 26065.033126033304
Residual deviance: 17375.72953938096
AIC: 17459.72953938096
AUC: 0.9070133046414631
pr_auc: 0.5455021097651589
Gini: 0.8140266092829263
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.21923662065754926: 


no     yes    Error    Rate
-----  -----  -----  -------  ----------------
no     29524  2482   0.0775   (2482.0/32006.0)
yes    1509   2709   0.3578   (1509.0/4218.0)
Total  31033  5191   0.1102   (3991.0/36224.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.219237     0.575832  224
max f2                       0.105018     0.6888    288
max f0point5                 0.34658      0.571974  174
max accuracy                 0.402298     0.903158  155
max precision                0.906216     0.715493  23
max recall                   0.00312165   1         397
max specificity              0.999266     0.999625  0
max absolute_mcc             0.158954     0.52064   253
max min_per_class_accuracy   0.116856     0.834282  280
max mean_per_class_accuracy  0.103873     0.839231  289

Gains/Lift Table: Avg response rate: 11.64 %, avg score: 11.64 %



group    cumulative_data_fraction    lower_threshold    lift       cumulative_lift    response_rate    score       cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ---------  -----------------  ---------------  ----------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------
    1        0.010021                    0.90246            6.1275     6.1275             0.713499         0.954618    0.713499                    0.954618            0.0614035       0.0614035                  512.75    512.75
    2        0.0200144                   0.809957           5.93091    6.02934            0.690608         0.852681    0.702069                    0.90372             0.0592698       0.120673                   493.091   502.934
    3        0.0300077                   0.720727           5.85974    5.97286            0.68232          0.762232    0.695492                    0.856601            0.0585586       0.179232                   485.974   497.286
    4        0.0400011                   0.632535           5.66995    5.89718            0.660221         0.675831    0.68668                     0.811439            0.0566619       0.235894                   466.995   489.718
    5        0.0500221                   0.561567           4.9919     5.71583            0.581267         0.594942    0.665563                    0.768068            0.0500237       0.285917                   399.19    471.583
    6        0.100017                    0.316307           4.358      5.0371             0.507454         0.42414     0.58653                     0.596151            0.217876        0.503793                   335.8     403.71
    7        0.150011                    0.206682           3.0634     4.37932            0.356709         0.255844    0.509937                    0.482736            0.153153        0.656946                   206.34    337.932
    8        0.200006                    0.14674            2.26199    3.85006            0.26339          0.172992    0.448309                    0.405311            0.113087        0.770033                   126.199   285.006
    9        0.299994                    0.088061           1.23532    2.97856            0.143843         0.113318    0.34683                     0.307989            0.123518        0.893551                   23.5319   197.856
    10       0.400011                    0.0616198          0.493043   2.35709            0.057411         0.0732211   0.274465                    0.249289            0.0493125       0.942864                   -50.6957  135.709
    11       0.5                         0.0463588          0.258445   1.93741            0.0300939        0.0534209   0.225596                    0.21012             0.0258416       0.968706                   -74.1555  93.7411
    12       0.599989                    0.0350418          0.147005   1.63904            0.0171176        0.040577    0.190853                    0.181865            0.0146989       0.983404                   -85.2995  63.9038
    13       0.700006                    0.0259352          0.0853344  1.41704            0.00993652       0.0303479   0.165004                    0.160216            0.00853485      0.991939                   -91.4666  41.7045
    14       0.799994                    0.0176356          0.0284526  1.24349            0.00331309       0.0217306   0.144795                    0.142907            0.00284495      0.994784                   -97.1547  24.3489
    15       0.899983                    0.0103586          0.0284526  1.1085             0.00331309       0.0138658   0.129076                    0.128571            0.00284495      0.997629                   -97.1547  10.8497
    16       1                           5.20281e-05        0.023704   1                  0.00276014       0.006



ModelMetricsBinomialGLM: glm
** Reported on validation data. **

MSE: 0.0706300616089855
RMSE: 0.2657631682701452
LogLoss: 0.23828595057289353
Null degrees of freedom: 4594
Residual degrees of freedom: 4553
Null deviance: 3362.8735995876823
Residual deviance: 2189.847885764891
AIC: 2273.847885764891
AUC: 0.9126833311273722
pr_auc: 0.5596829193662611
Gini: 0.8253666622547444
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.20746383675718202: 


no    yes    Error    Rate
-----  ----  -----  -------  --------------
no     3702  344    0.085    (344.0/4046.0)
yes    169   380    0.3078   (169.0/549.0)
Total  3871  724    0.1116   (513.0/4595.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.207464     0.597015  214
max f2                       0.125146     0.708601  264
max f0point5                 0.330597     0.605758  164
max accuracy                 0.345432     0.906202  159
max precision                0.986044     0.733333  3
max recall                   0.0183376    1         376
max specificity              0.999115     0.999506  0
max absolute_mcc             0.159438     0.543964  241
max min_per_class_accuracy   0.125146     0.843352  264
max mean_per_class_accuracy  0.103834     0.846804  280

Gains/Lift Table: Avg response rate: 11.95 %, avg score: 12.19 %



group    cumulative_data_fraction    lower_threshold    lift       cumulative_lift    response_rate    score       cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ---------  -----------------  ---------------  ----------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------
    1        0.0100109                   0.907156           5.64049    5.64049            0.673913         0.964675    0.673913                    0.964675            0.0564663       0.0564663                  464.049   464.049
    2        0.0200218                   0.827702           5.09464    5.36757            0.608696         0.872039    0.641304                    0.918357            0.0510018       0.107468                   409.464   436.757
    3        0.0300326                   0.744951           5.64049    5.45854            0.673913         0.783756    0.652174                    0.87349             0.0564663       0.163934                   464.049   445.854
    4        0.0400435                   0.669443           6.7322     5.77696            0.804348         0.713021    0.690217                    0.833373            0.0673953       0.23133                    573.22    477.696
    5        0.0500544                   0.613435           5.64049    5.74966            0.673913         0.639607    0.686957                    0.794619            0.0564663       0.287796                   464.049   474.966
    6        0.100109                    0.350258           4.73074    5.2402             0.565217         0.464248    0.626087                    0.629434            0.236794        0.52459                    373.074   424.02
    7        0.150163                    0.218273           2.83844    4.43961            0.33913          0.273641    0.530435                    0.510836            0.142077        0.666667                   183.844   343.961
    8        0.2                         0.153065           2.4488     3.94353            0.292576         0.18321     0.471164                    0.429197            0.12204         0.788707                   144.88    294.353
    9        0.300109                    0.0904129          1.1281     3.00437            0.134783         0.116933    0.358956                    0.325034            0.112933        0.901639                   12.8099   200.437
    10       0.4                         0.0628949          0.437635   2.36339            0.0522876        0.0751264   0.282372                    0.262625            0.0437158       0.945355                   -56.2365  136.339
    11       0.500109                    0.046525           0.181951   1.92672            0.0217391        0.0540121   0.2302                      0.220866            0.0182149       0.96357                    -81.8049  92.6721
    12       0.6                         0.0347454          0.218817   1.64238            0.0261438        0.0401877   0.196228                    0.190786            0.0218579       0.985428                   -78.1183  64.238
    13       0.699891                    0.0262076          0.0729391  1.41838            0.0087146        0.0304128   0.169465                    0.167897            0.00728597      0.992714                   -92.7061  41.8383
    14       0.8                         0.0181864          0.0727805  1.25               0.00869565       0.0220271   0.149347                    0.149643            0.00728597      1                          -92.7219  25
    15       0.899891                    0.0106852          0          1.11125            0                0.014389    0.132769                    0.134629            0               1                          -100      11.1245
    16       1                           0.000161535        0          1                  0                0.00722296


Scoring History: 


timestamp            duration    iterations    negative_log_likelihood    objective
--  -------------------  ----------  ------------  -------------------------  -----------
    2019-02-05 11:15:17  0.000 sec   0             13032.5                    0.359776
    2019-02-05 11:15:17  0.182 sec   1             9798.42                    0.274585
    2019-02-05 11:15:18  0.277 sec   2             9056.4                     0.252149
    2019-02-05 11:15:18  0.341 sec   3             8696.59                    0.242959
    2019-02-05 11:15:18  0.384 sec   4             8688.39                    0.242812
    2019-02-05 11:15:18  0.427 sec   5             8687.86                    0.242807

In [0]:
%memit model.train(x=feature_columns, y=label, training_frame=train, validation_frame=val)

glm Model Build progress: |███████████████████████████████████████████████| 100%
peak memory: 151.50 MiB, increment: 1.01 MiB


In [0]:
%time model.train(x=feature_columns, y=label, training_frame=train, validation_frame=val)

glm Model Build progress: |███████████████████████████████████████████████| 100%
CPU times: user 46.4 ms, sys: 3.6 ms, total: 50 ms
Wall time: 544 ms


In [0]:
%prun model.train(x=feature_columns, y=label, training_frame=train, validation_frame=val)

glm Model Build progress: |███████████████████████████████████████████████| 100%
 

**Results**

**Accuracy**: 0.9062
**AUC**: 0.91268
**AUC Precision Recall**: 0.5291479  
**Average Loss**: 0.40094  
**Loss**: 0.2657
**Precision**: 0.9563
**Recall**: 0.9149  


**%time**    
CPU times: user 46.4 ms, sys: 3.6 ms, total: 50 ms  
Wall time: 544 ms

**%prun**   
30356 function calls (28684 primitive calls) in 0.797 seconds

**%memit**   
peak memory: 151.50 MiB, increment: 1.01 MiB

## Random Forest

In [0]:
model = H2ORandomForestEstimator(model_id="rd_v1", ntrees=200)

In [0]:
model.train(x=feature_columns, y=label, training_frame=train, validation_frame=val)

drf Model Build progress: |███████████████████████████████████████████████| 100%


In [0]:
model

Model Details
H2ORandomForestEstimator :  Distributed Random Forest
Model Key:  rd_v1


ModelMetricsBinomial: drf
** Reported on train data. **

MSE: 0.06268215294274644
RMSE: 0.2503640408340352
LogLoss: 0.20027416620381755
Mean Per-Class Error: 0.1308599765566716
AUC: 0.9315084524958824
pr_auc: 0.6113683366816596
Gini: 0.8630169049917649
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.29482220603287906: 


no     yes    Error    Rate
-----  -----  -----  -------  ----------------
no     29138  2868   0.0896   (2868.0/32006.0)
yes    1019   3199   0.2416   (1019.0/4218.0)
Total  30157  6067   0.1073   (3887.0/36224.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.294822     0.622071  222
max f2                       0.159154     0.735509  281
max f0point5                 0.453812     0.601779  156
max accuracy                 0.460234     0.907768  153
max precision                0.969679     1         0
max recall                   4.18022e-05  1         399
max specificity              0.969679     1         0
max absolute_mcc             0.292252     0.574683  223
max min_per_class_accuracy   0.181404     0.862994  270
max mean_per_class_accuracy  0.116843     0.86914   302

Gains/Lift Table: Avg response rate: 11.64 %, avg score: 11.89 %



group    cumulative_data_fraction    lower_threshold    lift        cumulative_lift    response_rate    score        cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ----------  -----------------  ---------------  -----------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------
    1        0.010021                    0.798965           6.81359     6.81359            0.793388         0.85438      0.793388                    0.85438             0.0682788       0.0682788                  581.359   581.359
    2        0.0200144                   0.725514           6.26304     6.53869            0.729282         0.76027      0.761379                    0.80739             0.0625889       0.130868                   526.304   553.869
    3        0.0300077                   0.679966           5.83601     6.30468            0.679558         0.702389     0.734131                    0.772422            0.0583215       0.189189                   483.601   530.468
    4        0.0400011                   0.638372           5.69367     6.15204            0.662983         0.658057     0.716356                    0.74385             0.056899        0.246088                   469.367   515.204
    5        0.0500221                   0.604018           5.46506     6.01441            0.636364         0.620624     0.700331                    0.719164            0.0547653       0.300853                   446.506   501.441
    6        0.100017                    0.460596           4.65675     5.33577            0.542242         0.531035     0.621308                    0.625126            0.232812        0.533665                   365.675   433.577
    7        0.150011                    0.33605            3.457       4.70963            0.40254          0.397386     0.548399                    0.549226            0.172831        0.706496                   245.7     370.963
    8        0.200006                    0.221265           2.47538     4.15114            0.288239         0.27651      0.483368                    0.481057            0.123755        0.830251                   147.538   315.114
    9        0.299994                    0.0811368          1.10491     3.13583            0.128658         0.140805     0.365142                    0.36765             0.110479        0.94073                    10.4911   213.583
    10       0.400011                    0.0315618          0.348449    2.43888            0.0405741        0.0517373    0.283989                    0.288661            0.0348506       0.975581                   -65.1551  143.888
    11       0.5                         0.0147186          0.113811    1.97392            0.0132523        0.021447     0.229848                    0.235224            0.0113798       0.986961                   -88.6189  97.3921
    12       0.599989                    0.00444001         0.0829869   1.65879            0.00966317       0.00990639   0.193154                    0.197674            0.00829777      0.995258                   -91.7013  65.8795
    13       0.700006                    0.00128009         0.0213336   1.42483            0.00248413       0.00244081   0.165911                    0.16978             0.00213371      0.997392                   -97.8666  42.4835
    14       0.799994                    0.000369089        0.0237105   1.24971            0.00276091       0.000750691  0.145519                    0.148653            0.00237079      0.999763                   -97.6289  24.9712
    15       0.900066                    4.55615e-05        0           1.11077            0                0.000172366  0.12934                     0.132145            0               0.999763                   -100      11.0766
    16       1                           0                  0.00237236  1   



ModelMetricsBinomial: drf
** Reported on validation data. **

MSE: 0.06171774416418573
RMSE: 0.2484305620574605
LogLoss: 0.19631429532563424
Mean Per-Class Error: 0.1270671881738874
AUC: 0.9369707381506123
pr_auc: 0.6309145213992088
Gini: 0.8739414763012245
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.3218680327013135: 


no    yes    Error    Rate
-----  ----  -----  -------  --------------
no     3726  320    0.0791   (320.0/4046.0)
yes    128   421    0.2332   (128.0/549.0)
Total  3854  741    0.0975   (448.0/4595.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.321868     0.652713  193
max f2                       0.159111     0.745389  267
max f0point5                 0.432792     0.640784  146
max accuracy                 0.432792     0.914037  146
max precision                0.938431     1         0
max recall                   0.000923078  1         395
max specificity              0.938431     1         0
max absolute_mcc             0.321868     0.606555  193
max min_per_class_accuracy   0.187545     0.864558  251
max mean_per_class_accuracy  0.147354     0.872933  272

Gains/Lift Table: Avg response rate: 11.95 %, avg score: 12.16 %



group    cumulative_data_fraction    lower_threshold    lift       cumulative_lift    response_rate    score        cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ---------  -----------------  ---------------  -----------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------
    1        0.0100109                   0.791478           6.55025    6.55025            0.782609         0.840135     0.782609                    0.840135            0.0655738       0.0655738                  555.025   555.025
    2        0.0200218                   0.724598           6.0044     6.27732            0.717391         0.755053     0.75                        0.797594            0.0601093       0.125683                   500.44    527.732
    3        0.0300326                   0.673754           5.45854    6.0044             0.652174         0.695835     0.717391                    0.763674            0.0546448       0.180328                   445.854   500.44
    4        0.0400435                   0.637025           5.82244    5.95891            0.695652         0.655423     0.711957                    0.736611            0.0582878       0.238616                   482.244   495.891
    5        0.0500544                   0.608213           5.64049    5.89522            0.673913         0.624138     0.704348                    0.714117            0.0564663       0.295082                   464.049   489.522
    6        0.100109                    0.470397           5.05825    5.47674            0.604348         0.537378     0.654348                    0.625747            0.253188        0.54827                    405.825   447.674
    7        0.150163                    0.348811           3.56625    4.83991            0.426087         0.406276     0.578261                    0.55259             0.178506        0.726776                   256.625   383.991
    8        0.2                         0.226807           2.0833     4.15301            0.248908         0.288549     0.496192                    0.486795            0.103825        0.830601                   108.33    315.301
    9        0.300109                    0.085471           1.1281     3.14397            0.134783         0.147967     0.375635                    0.373771            0.112933        0.943534                   12.8099   214.397
    10       0.4                         0.0336135          0.455869   2.47268            0.0544662        0.0544704    0.29543                     0.294032            0.0455373       0.989071                   -54.4131  147.268
    11       0.500109                    0.0151002          0.0181951  1.98135            0.00217391       0.0226331    0.236728                    0.239705            0.00182149      0.990893                   -98.1805  98.1354
    12       0.6                         0.00667083         0.0182348  1.65452            0.00217865       0.0104558    0.197679                    0.201539            0.00182149      0.992714                   -98.1765  65.4523
    13       0.699891                    0.0030316          0.0364696  1.42359            0.0043573        0.00503904   0.170087                    0.173493            0.00364299      0.996357                   -96.353   42.3588
    14       0.8                         0.000832978        0.0363903  1.25               0.00434783       0.00166828   0.149347                    0.151992            0.00364299      1                          -96.361   25
    15       0.899891                    0.000197159        0          1.11125            0                0.000472716  0.132769                    0.135173            0               1                          -100      11.1245
    16       1                           0                  0          1                  0        


Scoring History: 


timestamp            duration    number_of_trees    training_rmse    training_logloss    training_auc    training_pr_auc    training_lift    training_classification_error    validation_rmse    validation_logloss    validation_auc    validation_pr_auc    validation_lift    validation_classification_error
--  -------------------  ----------  -----------------  ---------------  ------------------  --------------  -----------------  ---------------  -------------------------------  -----------------  --------------------  ----------------  -------------------  -----------------  ---------------------------------
    2019-02-05 11:23:06  0.032 sec   0                  nan              nan                 nan             nan                nan              nan                              nan                nan                   nan               nan                  nan                nan
    2019-02-05 11:23:07  0.780 sec   1                  0.339135         3.63773             0.68504         0.118396           4.30674          0.119558                         0.339371           3.62667               0.703526          0.121513             4.2775             0.119478
    2019-02-05 11:23:07  1.255 sec   2                  0.329503         3.27633             0.716863        0.165181           4.37731          0.120053                         0.30503            1.97677               0.785326          0.269553             4.99591            0.138847
    2019-02-05 11:23:08  1.639 sec   3                  0.324813         2.99494             0.733445        0.198348           4.38844          0.129582                         0.287301           1.10523               0.834703          0.373157             5.85203            0.143417
    2019-02-05 11:23:08  2.088 sec   4                  0.31641          2.63304             0.755764        0.232222           4.53567          0.129772                         0.279921           0.773353              0.865673          0.433458             5.82705            0.117084
    2019-02-05 11:23:09  2.591 sec   5                  0.308814         2.2593              0.774982        0.266309           4.62704          0.13409                          0.274888           0.58949               0.878539          0.468245             6.32003            0.114908
    2019-02-05 11:23:09  3.132 sec   6                  0.302821         1.97122             0.7922          0.297388           4.742            0.138187                         0.270019           0.492686              0.891392          0.504729             6.18635            0.118607
    2019-02-05 11:23:10  3.658 sec   7                  0.296894         1.69544             0.81017         0.326719           4.782            0.138764                         0.267471           0.401361              0.900528          0.523105             5.85883            0.124483
    2019-02-05 11:23:14  7.776 sec   23                 0.262547         0.381462            0.903175        0.529045           5.94388          0.114731                         0.253459           0.245423              0.925203          0.608435             6.59924            0.102285
    2019-02-05 11:23:18  11.881 sec  51                 0.254864         0.240426            0.922165        0.584798           6.50603          0.113488                         0.251147           0.207601              0.932401          0.615852             6.10295            0.0977149
    2019-02-05 11:23:22  15.885 sec  86                 0.252378         0.215077            0.927743        0.600539           6.72919          0.112136                         0.249416           0.204502              0.934902          0.627218             7.0961             0.0981502
    2019-02-05 11:23:26  19.902 sec  122                0.251335         0.202673            0.930392        0.604187           6.62432          0.109044                         0.249215           0.198229              0.935522          0.63

Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
duration    130268                 1                    0.287941
month       59591.2                0.45745              0.131718
age         41046.8                0.315095             0.0907286
day         38325.8                0.294207             0.0847142
poutcome    35658.2                0.273729             0.0788178
job         33843.3                0.259797             0.0748062
balance     24444.9                0.187651             0.0540322
pdays       19133.8                0.14688              0.0422928
campaign    14689.5                0.112764             0.0324693
education   13098.5                0.10055              0.0289524
marital     9867.99                0.0757513            0.0218119
housing     9584.6                 0.0735759            0.0211855
contact     9506.22                0.0729742            0.0210122
previous    8011.15                0.0614973            0.0177076
loan        4361.45                0.0334805            0.00964041
default     981.594                0.00753517           0.00216968

In [0]:
%memit model.train(x=feature_columns, y=label, training_frame=train, validation_frame=val)

drf Model Build progress: |███████████████████████████████████████████████| 100%
peak memory: 152.40 MiB, increment: 0.38 MiB


In [0]:
%time model.train(x=feature_columns, y=label, training_frame=train, validation_frame=val)

drf Model Build progress: |███████████████████████████████████████████████| 100%
CPU times: user 331 ms, sys: 47.6 ms, total: 379 ms
Wall time: 24.4 s


In [0]:
%prun model.train(x=feature_columns, y=label, training_frame=train, validation_frame=val)

drf Model Build progress: |███████████████████████████████████████████████| 100%
 

**Results**

**Train Loss (RMSE)**: 0.161641  
**Validation Loss(RMSE)**: 0.162671 

**%time**    
CPU times: user 331 ms, sys: 47.6 ms, total: 379 ms   
Wall time: 24.4 s

**%prun**   
211981 function calls (206241 primitive calls) in 24.111 seconds 

**%memit**   
peak memory: 152.40 MiB, increment: 0.38 MiB